### PR TITLE
Fix link account

### PR DESCRIPTION
**Pull request by bab29**
Updated to allow for 11.7- to work properly

### CODE REVIEW

File: Linked Accounts/Link-Accounts.ps1 [hunk 1]
```diff
@@ -67,6 +64,7 @@ $URL_Logoff = $URL_Authentication + "/Logoff"
 # -----------
 $URL_Accounts = $URL_PVWAAPI + "/Accounts"
 $URL_LinkAccounts = $URL_PVWAAPI + "/Accounts/{0}/LinkAccount"
+$URL_Version = $URL_PIMServer + "/Server"
```
Comment (AssafMiron) [inline Linked Accounts/Link-Accounts.ps1:67]: @bab29 you have the same functionality also in /api/sever
Don't think you need to add the WebServices URL for that

File: Linked Accounts/Link-Accounts.ps1 [hunk 2]
```diff
@@ -108,37 +98,49 @@ Function ConvertTo-URL($sText)
...
 		Write-Debug "Returning URL Encode of $sText"
 		return [URI]::EscapeDataString($sText)
-	}
-	else
-	{
+	} else {
 		return $sText
 	}
 }
+Function Compare-Version () {
```
Comment (AssafMiron) [inline Linked Accounts/Link-Accounts.ps1:108]: @bab29 , Would you consider using a method I built for this exact use case?

# @FUNCTION@ ======================================================================================================================
# Name...........: Test-RESTVersion
# Description....: Tests if the requested version exists in the PVWA REST API
# Parameters.....: version
# Return Values..: True / False
# =================================================================================================================================
Function Test-RESTVersion
{
<#
.SYNOPSIS
	Tests if the requested version exists in the PVWA REST API
.DESCRIPTION
	Tests if the requested version exists in the PVWA REST API
.PARAMETER Version
	A string of the requested PVWA REST version to test
#>

    param (
		[Parameter(Mandatory=$true)]
		[string]$version
	)

	$retVersionExists = $false
	try{
		Write-LogMessage -Type debug -Msg "Testing to see if the PVWA is at least in version $version"
		$serverResponse = Invoke-REST -Command GET -URI $URL_Server
		if($null -ne $serverResponse)
		{
			Write-LogMessage -Type debug -Msg "The current PVWA is in version $($serverResponse.ExternalVersion)"
			$majorVersionCompare = ( [int]($serverResponse.ExternalVersion.Split('.')[0]) -gt [int]($version.Split('.')[0]) )
			$majorVersionEqual = ( [int]($serverResponse.ExternalVersion.Split('.')[0]) -eq [int]($version.Split('.')[0]) )
			$minorVersionCompare = ( [int]($serverResponse.ExternalVersion.Split('.')[1]) -ge [int]($version.Split('.')[1]) )
			if($majorVersionCompare -or ($majorVersionEqual -and $minorVersionCompare))
			{
				$retVersionExists = $true
			}
		}
		else
		{
			Throw "An error occurred while testing the PVWA version"
		}

		return $retVersionExists
	} catch {
		# Check the error code returned from the REST call
		$innerExcp = $_.Exception.InnerException
		Write-LogMessage -Type Verbose -Msg "Status Code: $($innerExcp.StatusCode); Status Description: $($innerExcp.StatusDescription); REST Error: $($innerExcp.CyberArkErrorMessage)"
		if($innerExcp.StatusCode -eq "NotFound") {
			return $false
		}
		else{
			Throw $(New-Object System.Exception ("Test-RESTVersion: There was an error checking for REST version $version.",$_.Exception))
		}
	}
}

Comment (AssafMiron) [reply Linked Accounts/Link-Accounts.ps1:108]: You use it the same way:
If (Test-RESTVersion -version "10.4")
{...}

File: Linked Accounts/Link-Accounts.ps1 [hunk 3]
```diff
@@ -613,98 +528,94 @@ Write-LogMessage -Type Info -MSG "Getting PVWA Credentials to start Linking acco
...
 $delimiter = $((Get-Culture).TextInfo.ListSeparator)
 $accountsCSV = Import-Csv $csvPath -Delimiter $delimiter
-$rowCount = $accountsCSV.Count
-$counterMaster = 0
-Write-LogMessage -Type Info -MSG "Starting to add links to $rowCount master accounts" -SubHeader
+
+$ExtraPass1Count = @($accountsCSV |Where-Object ExtraPass1Name -ne "" ).count
+$ExtraPass2Count = @($accountsCSV |Where-Object ExtraPass2Name -ne "" ).count
+$ExtraPass3Count = @($accountsCSV |Where-Object ExtraPass3Name -ne "" ).count
+$linkCount =  $ExtraPass1Count + $ExtraPass2Count + $ExtraPass3Count
```
Comment (AssafMiron) [inline Linked Accounts/Link-Accounts.ps1:555]: What is this used for?
I didn't find any use for this counter in your code

File: Linked Accounts/Link-Accounts.ps1 [hunk 4]
```diff
@@ -613,98 +528,94 @@ Write-LogMessage -Type Info -MSG "Getting PVWA Credentials to start Linking acco
...
+
+$masterCount = @($accountsCSV | Select-Object -Property userName,address -Unique).Count
```
Comment (AssafMiron) [inline Linked Accounts/Link-Accounts.ps1:557]: Same here, what do you need this counter for?

File: Linked Accounts/Link-Accounts.ps1 [hunk 5]
```diff
@@ -613,98 +528,94 @@ Write-LogMessage -Type Info -MSG "Getting PVWA Credentials to start Linking acco
...
-}
-else
-{
+} else {
 	Write-LogMessage -Type Error -MSG "No Credentials were entered" -Footer
 	exit
 }
 #endregion
 
+If (Compare-Version("11.7")) {$extraPass = "extraPasswordIndex"} else {$extraPass = "extraPasswordID"}
```
Comment (AssafMiron) [inline Linked Accounts/Link-Accounts.ps1:543]: @bab29 from this If I understand that there was a change in the API parameter name
so this means that we need to check a minimum version (that has the LinkAccounts API URL) and a change version (from which we had the parameter change)
I don't think a simple If/Else is enough here

WDYT?

**Review by AssafMiron — CHANGES_REQUESTED** [review]
Hey @bab29 
I added a few more comments
Thanks for the effort on this

File: Linked Accounts/Link-Accounts.ps1 (hunk 1 above)
Comment (bab29) [reply Linked Accounts/Link-Accounts.ps1:67]: Using the PIMServers.svc is what is in documentation. At what point was it added to /api/server?

File: Linked Accounts/Link-Accounts.ps1 (hunk 2 above)
Comment (bab29) [reply Linked Accounts/Link-Accounts.ps1:108]: Sure, I will make this update.

File: Linked Accounts/Link-Accounts.ps1 (hunk 4 above)
Comment (bab29) [reply Linked Accounts/Link-Accounts.ps1:557]: It's used on line 631

**Comment by AssafMiron** [comment]
Long time ago, I guess it was never documented
But it is there

**Comment by bab29** [comment]
Ok, I can make that update. Any clue around what version it was added?

***@***.***
Brian Andrew Bors
Sr. Security Services Engineer
T: +617 965 1544 x 3984    |  M: + 509 655 8090
***@***.******@***.***>  |  www.cyberark.com<http://www.cyberark.com/>

File: Linked Accounts/Link-Accounts.ps1 (hunk 1 above)
Comment (AssafMiron) [reply Linked Accounts/Link-Accounts.ps1:67]: 10.3 if I'm not mistaken

File: Linked Accounts/Link-Accounts.ps1 (hunk 2 above)
Comment (bab29) [reply Linked Accounts/Link-Accounts.ps1:108]: I added this but isn't this easier then doing all the splits adn compare?

If ([version]($serverResponse.InternalVersion) -ge [version]$version) {$retVersionExists = $true}

**Review by AssafMiron — APPROVED** [review]

**Comment by AssafMiron** [comment]
@bab29  looks good
Let me know if you are ready to merge or there are more changes pending

**Comment by bab29** [comment]
It is ready to go.

**Pull request merged**